aiida/tools/importexport/dbexport/__init__.py

```diff
@@ -7,81 +7,293 @@
-# pylint: disable=fixme,too-many-branches,too-many-locals,too-many-statements,too-many-arguments
+# pylint: disable=fixme,too-many-lines
 """Provides export functionalities."""
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from dataclasses import dataclass
 import logging
 import os
 import tarfile
 import time
+from typing import (
+    Any,
+    Callable,
+    DefaultDict,
+    Dict,
+    Iterable,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Type,
+    Union,
+    cast,
+)
 
 from aiida import get_version, orm
 from aiida.common import json
 from aiida.common.exceptions import LicensingException
-from aiida.common.folders import RepositoryFolder, SandboxFolder, Folder
+from aiida.common.folders import Folder, RepositoryFolder, SandboxFolder
 from aiida.common.lang import type_check
-from aiida.common.log import override_log_formatter, LOG_LEVEL_REPORT
+from aiida.common.log import LOG_LEVEL_REPORT, override_log_formatter
 from aiida.orm.utils._repository import Repository
-
-from aiida.tools.importexport.common import exceptions, get_progress_bar, close_progress_bar
-from aiida.tools.importexport.common.config import EXPORT_VERSION, NODES_EXPORT_SUBFOLDER
-from aiida.tools.importexport.common.config import (
-    NODE_ENTITY_NAME, GROUP_ENTITY_NAME, COMPUTER_ENTITY_NAME, LOG_ENTITY_NAME, COMMENT_ENTITY_NAME
+from aiida.tools.importexport.common import (
+    close_progress_bar,
+    exceptions,
+    get_progress_bar,
 )
 from aiida.tools.importexport.common.config import (
-    get_all_fields_info, file_fields_to_model_fields, entity_names_to_entities, model_fields_to_file_fields
+    COMMENT_ENTITY_NAME,
+    COMPUTER_ENTITY_NAME,
+    EXPORT_VERSION,
+    GROUP_ENTITY_NAME,
+    LOG_ENTITY_NAME,
+    NODE_ENTITY_NAME,
+    NODES_EXPORT_SUBFOLDER,
+    entity_names_to_entities,
+    file_fields_to_model_fields,
+    get_all_fields_info,
+    model_fields_to_file_fields,
 )
+from aiida.tools.graph.graph_traversers import get_nodes_export
 from aiida.tools.importexport.common.utils import export_shard_uuid
 from aiida.tools.importexport.dbexport.utils import (
-    check_licenses, fill_in_query, serialize_dict, check_process_nodes_sealed, summary, EXPORT_LOGGER, ExportFileFormat,
-    deprecated_parameters
+    EXPORT_LOGGER,
+    ExportFileFormat,
+    check_licenses,
+    check_process_nodes_sealed,
+    deprecated_parameters,
+    fill_in_query,
+    serialize_dict,
+    summary,
 )
-
-from .zip import ZipFolder
+from aiida.tools.importexport.dbexport.zip import ZipFolder
 
 __all__ = ('export', 'EXPORT_LOGGER', 'ExportFileFormat')
 
 
+@dataclass
+class ExportData:
+    """Class for storing data to export."""
+    metadata: dict
+    node_uuids: Set[str]
+    group_uuids: Dict[str, List[str]]
+    link_uuids: List[dict]
+    # all entity data from the database, except Node extras and attributes
+    entity_data: Dict[str, Dict[int, dict]]
```

Comment:
well I've just initially copied what was being used previously.
It certainly could be though.
node_data was the main one I wanted to make iterable, because that will be the one that takes up the most memory. I actually want to make it a generator, but just need to make sure that works with the progress bar updates